File: android/sdk/src/main/kotlin/com/ary/push/internal/sync/OperationQueue.kt
package com.ary.push.internal.sync

import com.ary.push.internal.PushJson
import com.ary.push.internal.log.PushLogger
import com.ary.push.internal.storage.StorageManager
import org.json.JSONArray

/**
 * The durable, bounded, self-coalescing queue behind offline support.
 *
 * Three properties matter:
 *
 *  * **Durable.** Every mutation is committed synchronously, so a `login()` made while offline is
 *    still pending after the process is killed and the device rebooted.
 *  * **Coalescing.** Enqueueing collapses redundant work at the source rather than sending it and
 *    hoping the backend copes: three `addTag` calls become one PATCH, and a newer token replaces
 *    an older unsent one instead of queueing behind it.
 *  * **Bounded.** A device that is offline for a week must not accumulate unbounded state.
 *    Beyond [MAX_OPERATIONS] the oldest low-value entries (events) are dropped first.
 *
 * Access is synchronised on the instance; the queue is small and contention is negligible.
 */
internal class OperationQueue(private val storage: StorageManager) {

    private val lock = Any()

    /** Adds an operation, collapsing it into an equivalent pending one where possible. */
    fun enqueue(operation: PendingOperation): Unit = synchronized(lock) {
        val current = readInternal().toMutableList()

        when {
            operation.type.isLatestValueWins ->
                current.removeAll { it.type == operation.type }

            operation.type == OperationType.UPDATE_TAGS -> {
                // Merge into the pending tag write so a burst of addTag calls costs one request.
                val existingIndex = current.indexOfFirst { it.type == OperationType.UPDATE_TAGS }
                if (existingIndex >= 0) {
                    val merged = mergeTagPayloads(current[existingIndex], operation)
                    current[existingIndex] = merged
                    writeInternal(current)
                    return
                }
            }
        }

        // A logout supersedes an unsent identify, and vice versa: sending both is contradictory.
        when (operation.type) {
            OperationType.LOGOUT_USER -> current.removeAll { it.type == OperationType.IDENTIFY_USER }
            OperationType.IDENTIFY_USER -> current.removeAll { it.type == OperationType.LOGOUT_USER }
            else -> Unit
        }

        current += operation
        writeInternal(current.enforceBound())
    }

    /** Snapshot of pending operations in dependency order. */
    fun snapshot(): List<PendingOperation> = synchronized(lock) {
        readInternal().sortedWith(compareBy({ it.type.ordinal }, { it.createdAt }))
    }

    fun isEmpty(): Boolean = synchronized(lock) { readInternal().isEmpty() }

    fun size(): Int = synchronized(lock) { readInternal().size }

    /** Removes a completed or permanently failed operation. */
    fun remove(operationId: String): Unit = synchronized(lock) {
        writeInternal(readInternal().filterNot { it.id == operationId })
    }

    /** Records a failed attempt so that backoff and the permanent-failure cut-off survive restarts. */
    fun recordAttempt(operationId: String): Unit = synchronized(lock) {
        writeInternal(
            readInternal().map {
                if (it.id == operationId) it.copy(attempts = it.attempts + 1) else it
            }
        )
    }

    fun clear(): Unit = synchronized(lock) {
        storage.putString(StorageManager.KEY_PENDING_OPERATIONS, null, durable = true)
    }

    // ------------------------------------------------------------------ persistence

    private fun readInternal(): List<PendingOperation> {
        val raw = storage.getString(StorageManager.KEY_PENDING_OPERATIONS) ?: return emptyList()
        return try {
            val array = JSONArray(raw)
            buildList(array.length()) {
                for (index in 0 until array.length()) {
                    val element = array.optJSONObject(index) ?: continue
                    PendingOperation.fromJson(element)?.let(::add)
                }
            }
        } catch (t: Throwable) {
            // A corrupted queue is dropped rather than retried forever. Losing deferred
            // synchronisation is recoverable; a permanently wedged queue is not.
            PushLogger.w(t) { "Pending operation queue was unreadable and has been reset" }
            storage.putString(StorageManager.KEY_PENDING_OPERATIONS, null, durable = true)
            emptyList()
        }
    }

    private fun writeInternal(operations: List<PendingOperation>) {
        if (operations.isEmpty()) {
            storage.putString(StorageManager.KEY_PENDING_OPERATIONS, null, durable = true)
            return
        }
        val array = JSONArray()
        operations.forEach { array.put(it.toJson()) }
        storage.putString(StorageManager.KEY_PENDING_OPERATIONS, array.toString(), durable = true)
    }

    private fun mergeTagPayloads(
        existing: PendingOperation,
        incoming: PendingOperation
    ): PendingOperation {
        val existingTags = PushJson.flattenToStringMap(
            PushJson.parseObject(existing.payload[PendingOperation.KEY_TAGS])
        )
        val incomingTags = PushJson.flattenToStringMap(
            PushJson.parseObject(incoming.payload[PendingOperation.KEY_TAGS])
        )
        val merged = existingTags + incomingTags
        return existing.copy(
            payload = mapOf(
                PendingOperation.KEY_TAGS to PushJson.toJsonObject(merged).toString()
            ),
            // Merging produces a new value, so the attempt counter starts over.
            attempts = 0
        )
    }

    private fun List<PendingOperation>.enforceBound(): List<PendingOperation> {
        if (size <= MAX_OPERATIONS) return this
        PushLogger.w { "Pending operation queue exceeded $MAX_OPERATIONS entries; trimming" }
        val (events, rest) = partition { it.type == OperationType.TRACK_EVENTS }
        val trimmedEvents = events.sortedBy { it.createdAt }
            .drop((size - MAX_OPERATIONS).coerceAtMost(events.size))
        val combined = rest + trimmedEvents
        return if (combined.size <= MAX_OPERATIONS) {
            combined
        } else {
            combined.sortedBy { it.createdAt }.takeLast(MAX_OPERATIONS)
        }
    }

    private companion object {
        /** Deliberately small: this is deferred bookkeeping, not a message store. */
        const val MAX_OPERATIONS = 100
    }
}

File: ios/Sources/ARYPush/Internal/OperationQueue.swift
import Foundation

/// The durable, bounded, self-coalescing queue behind offline support.
///
/// Three properties matter:
///
/// * **Durable.** Every mutation is written straight through, so a `login()` made in aeroplane
///   mode is still pending after the app is killed and the device rebooted.
/// * **Coalescing.** Enqueueing collapses redundant work at the source rather than sending it
///   and hoping the backend copes: three `addTag` calls become one PATCH, and a newer token
///   replaces an older unsent one instead of queueing behind it.
/// * **Bounded.** A device offline for a week must not accumulate unbounded state. Beyond
///   `maxOperations` the oldest low-value entries (events) are dropped first.
final class PendingOperationQueue {

    /// Deliberately small: this is deferred bookkeeping, not a message store.
    private static let maxOperations = 100

    private let storage: StorageManager
    private let lock = NSRecursiveLock()

    init(storage: StorageManager) {
        self.storage = storage
    }

    /// Adds an operation, collapsing it into an equivalent pending one where possible.
    func enqueue(_ operation: PendingOperation) {
        lock.lock()
        defer { lock.unlock() }

        var current = read()

        if operation.type.isLatestValueWins {
            current.removeAll { $0.type == operation.type }
        } else if operation.type == .updateTags,
                  let index = current.firstIndex(where: { $0.type == .updateTags }) {
            // Merge into the pending tag write so a burst of addTag calls costs one request.
            current[index] = merged(existing: current[index], incoming: operation)
            write(current)
            return
        }

        // A logout supersedes an unsent identify, and vice versa: sending both is contradictory.
        switch operation.type {
        case .logoutUser: current.removeAll { $0.type == .identifyUser }
        case .identifyUser: current.removeAll { $0.type == .logoutUser }
        default: break
        }

        current.append(operation)
        write(bounded(current))
    }

    /// Snapshot of pending operations in dependency order.
    func snapshot() -> [PendingOperation] {
        lock.lock()
        defer { lock.unlock() }
        return read().sorted {
            $0.type.order == $1.type.order
                ? $0.createdAt < $1.createdAt
                : $0.type.order < $1.type.order
        }
    }

    var isEmpty: Bool {
        lock.lock(); defer { lock.unlock() }
        return read().isEmpty
    }

    var count: Int {
        lock.lock(); defer { lock.unlock() }
        return read().count
    }

    /// Removes a completed or permanently failed operation.
    func remove(id: String) {
        lock.lock(); defer { lock.unlock() }
        write(read().filter { $0.id != id })
    }

    /// Records a failed attempt so backoff and the permanent-failure cut-off survive restarts.
    func recordAttempt(id: String) {
        lock.lock(); defer { lock.unlock() }
        write(read().map { $0.id == id ? $0.withAttempts($0.attempts + 1) : $0 })
    }

    func clear() {
        lock.lock(); defer { lock.unlock() }
        storage.set(StorageManager.Keys.pendingOperations, nil)
    }

    // MARK: - Persistence

    private func read() -> [PendingOperation] {
        guard let raw = storage.string(StorageManager.Keys.pendingOperations),
              let data = raw.data(using: .utf8)
        else { return [] }

        guard let array = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            // A corrupted queue is dropped rather than retried forever. Losing deferred
            // synchronisation is recoverable; a permanently wedged queue is not.
            PushLogger.warn("Pending operation queue was unreadable and has been reset")
            storage.set(StorageManager.Keys.pendingOperations, nil)
            return []
        }
        return array.compactMap(PendingOperation.from(dictionary:))
    }

    private func write(_ operations: [PendingOperation]) {
        guard !operations.isEmpty else {
            storage.set(StorageManager.Keys.pendingOperations, nil)
            return
        }
        let payload = operations.map { $0.toDictionary() }
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let encoded = String(data: data, encoding: .utf8)
        else {
            PushLogger.warn("Could not persist the pending operation queue")
            return
        }
        storage.set(StorageManager.Keys.pendingOperations, encoded)
    }

    private func merged(
        existing: PendingOperation,
        incoming: PendingOperation
    ) -> PendingOperation {
        var tags = decodeTags(existing.payload[PendingOperation.Key.tags])
        decodeTags(incoming.payload[PendingOperation.Key.tags]).forEach { tags[$0.key] = $0.value }
        // Merging produces a new value, so the attempt counter starts over.
        return existing.withPayload([PendingOperation.Key.tags: encodeTags(tags)], attempts: 0)
    }

    private func bounded(_ operations: [PendingOperation]) -> [PendingOperation] {
        guard operations.count > Self.maxOperations else { return operations }
        PushLogger.warn(
            "Pending operation queue exceeded \(Self.maxOperations) entries; trimming"
        )
        let overflow = operations.count - Self.maxOperations
        let events = operations.filter { $0.type == .trackEvents }.sorted { $0.createdAt < $1.createdAt }
        let dropped = Set(events.prefix(min(overflow, events.count)).map(\.id))
        let kept = operations.filter { !dropped.contains($0.id) }
        guard kept.count > Self.maxOperations else { return kept }
        return Array(kept.sorted { $0.createdAt < $1.createdAt }.suffix(Self.maxOperations))
    }

    func decodeTags(_ raw: String?) -> [String: String] {
        guard let raw, let data = raw.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: String]
        else { return [:] }
        return json
    }

    func encodeTags(_ tags: [String: String]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: tags),
              let encoded = String(data: data, encoding: .utf8)
        else { return "{}" }
        return encoded
    }
}
